#Nama : Nabila Sukmanegara


# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [22]:
import pandas as pd
import numpy as np

## Input Dataset

In [23]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [24]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [25]:
#Menampilkan data
df.head(10)

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0          1    15634602    France  Female   42       2       0.00   
1          2    15647311     Spain  Female   41       1   83807.86   
2          3    15619304    France  Female   42       8  159660.80   
3          4    15701354    France  Female   39       1       0.00   
4          5    15737888     Spain  Female   43       2  125510.82   
5          6    15574012     Spain    Male   44       8  113755.78   
6          7    15592531    France    Male   50       7       0.00   
7          8    15656148   Germany  Female   29       4  115046.74   
8          9    15792365    France    Male   44       4  142051.07   
9         10    15592389    France    Male   27       2  134603.88   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0  
5              2          1               0        149756.71       1  
6              2          1               1         10062.80       0  
7              4          1               0        119346.88       1  
8              2          0               1         74940.50       0  
9              1          1               1         71725.73       0

In [26]:
#menampilkan ukuran data
df.shape

(10000, 12)

In [27]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
# Hilangkan kolom 'RowNumber' dan pilih kolom-kolom yang relevan
df = df[['Geography', 'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary', 'Exited']].copy()
df.head()

Geography  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0    France  Female   42       2       0.00              1          1   
1     Spain  Female   41       1   83807.86              1          0   
2    France  Female   42       8  159660.80              3          1   
3    France  Female   39       1       0.00              2          0   
4     Spain  Female   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  
0               1        101348.88       1  
1               1        112542.58       0  
2               0        113931.57       1  
3               0         93826.63       0  
4               1         79084.10       0

## Preprocessing

In [28]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies
# Menggunakan get_dummies untuk melakukan One-Hot Encoding pada kolom-kolom kategori
df_encoded = pd.get_dummies(df, columns=['Geography', 'Gender'])

# Menampilkan 5 baris pertama data yang telah diencode
df_encoded.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Exited  Geography_France  Geography_Germany  \
0        101348.88       1                 1                  0   
1        112542.58       0                 0                  0   
2        113931.57       1                 1                  0   
3         93826.63       0                 1                  0   
4         79084.10       0                 0                  0   

   Geography_Spain  Gender_Female  Gender_Male  
0                0              1            0  
1                1              1            0  
2                0              1            0  
3                0              1            0  
4                1              1            0

In [29]:
#Pisahkan variabel fitur (X) dan variabel target (Y)

#X berisi semua kolom kecuali "Exited"
X = df_encoded.drop(columns=['Exited'])
#Y berisi hanya kolom "Exited"
y = df_encoded['Exited']

In [30]:
#Menampilkan beberapa contoh data fitur (X)
X.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Geography_France  Geography_Germany  Geography_Spain  \
0        101348.88                 1                  0                0   
1        112542.58                 0                  0                1   
2        113931.57                 1                  0                0   
3         93826.63                 1                  0                0   
4         79084.10                 0                  0                1   

   Gender_Female  Gender_Male  
0              1            0  
1              1            0  
2              1            0  
3              1            0  
4              1            0

In [31]:
#Menampilkan conton data target (y)
y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [33]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

# Inisialisasi MinMaxScaler
scaler = MinMaxScaler()

# Melakukan normalisasi pada fitur (X_encoded)
X_transform = scaler.fit_transform(X)

# X_transform adalah hasil fitur yang telah dinormalisasi

## Train-Test Split

In [34]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

In [35]:
# Import library scikit-learn
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

# Inisialisasi model Decision Tree
model1 = DecisionTreeClassifier(random_state=42)  # Tambahkan random_state untuk hasil yang dapat direproduksi

# Definisikan daftar hyperparameter yang akan diuji
params = {
    "criterion": ["gini", "entropy"],
    "max_depth": [None, 10, 20, 30],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4]
}

# Buat objek GridSearchCV untuk mencari parameter terbaik
grid = GridSearchCV(
    estimator=model1,
    param_grid=params,
    scoring='accuracy',
    n_jobs=-1,  # Menggunakan semua core CPU yang tersedia
    cv=10
)

# Melakukan fitting model dengan data training
grid.fit(X_train, y_train)

# Menampilkan parameter terbaik yang ditemukan
best_params = grid.best_params_
print("Parameter terbaik untuk Decision Tree:", best_params)

Parameter terbaik untuk Decision Tree: {'criterion': 'entropy', 'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 10}


In [36]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.94      0.91      1983
           1       0.69      0.49      0.57       517

    accuracy                           0.85      2500
   macro avg       0.78      0.72      0.74      2500
weighted avg       0.84      0.85      0.84      2500


[[1871  112]
 [ 265  252]]

0.8492


**Membaca hasil output :**
- Precision:

Kelas 0 (Tidak Ada): Dari semua prediksi kelas 0, sekitar 88% benar (precision 0.89).
Kelas 1 (Ada): Dari semua prediksi kelas 1, sekitar 69% benar (precision 0.69).

- Recall:

Kelas 0 (Tidak Ada): Dari semua instance yang sebenarnya termasuk dalam kelas 0, model dapat memprediksi sekitar 94% dengan benar (recall 0.94).
Kelas 1 (Ada): Dari semua instance yang sebenarnya termasuk dalam kelas 1, model hanya dapat mendeteksi sekitar 49% (recall 0.49).

- F1-Score:

F1-score mencoba mencari keseimbangan antara precision dan recall. Untuk kelas 0, f1-score adalah 0.91, sedangkan untuk kelas 1, f1-score adalah 0.57.

- Matriks Konfusi:

Model berhasil memprediksi sebagian besar kelas 0 (1871 prediksi benar, 112 prediksi salah).
Model memiliki kesulitan dalam memprediksi kelas 1 (265 prediksi benar, 252 prediksi salah).

- Akurasi:

Akurasi model adalah 0.8492, yang menunjukkan bahwa model ini benar sekitar 84.92% dari waktu.

## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

In [37]:
# Import library
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

# Inisialisasi model SVM
model2 = SVC()

# Tentukan hyperparameter yang akan di-tune
param_grid = {
    # Parameter penalti untuk mengendalikan toleransi kesalahan dalam klasifikasi
    'C': [0.1, 1, 10],
    # Jenis kernel dalam SVM
    'kernel': ['linear', 'rbf'],
    # Parameter gamma (kernel rbf)
    'gamma': ['scale', 'auto', 0.1, 1]
}

# Buat objek GridSearchCV
grid = GridSearchCV(
    estimator=model2,
    param_grid=param_grid,
    scoring='accuracy',
    n_jobs=10,
    cv=10
)

# Fit model dan temukan parameter terbaik
grid.fit(X_train, y_train)
best_params = grid.best_params_
print("Parameter terbaik untuk SVM:", best_params)


Parameter terbaik untuk SVM: {'C': 10, 'gamma': 1, 'kernel': 'rbf'}


In [38]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.79      0.43      0.55       517

    accuracy                           0.86      2500
   macro avg       0.83      0.70      0.74      2500
weighted avg       0.85      0.86      0.84      2500


[[1924   59]
 [ 296  221]]

0.858


**Membaca Hasil output :**
- Precision:

Kelas 0 (Tidak Ada): Dari semua prediksi kelas 0, sekitar 87% benar (precision 0.87).
Kelas 1 (Ada): Dari semua prediksi kelas 1, sekitar 79% benar (precision 0.79).
- Recall:

Kelas 0 (Tidak Ada): Dari semua instance yang sebenarnya termasuk dalam kelas 0, model dapat memprediksi sekitar 97% dengan benar (recall 0.97).
Kelas 1 (Ada): Dari semua instance yang sebenarnya termasuk dalam kelas 1, model hanya dapat mendeteksi sekitar 43% (recall 0.43).
- F1-Score:

F1-score mencoba mencari keseimbangan antara precision dan recall. Untuk kelas 0, f1-score adalah 0.92, sedangkan untuk kelas 1, f1-score adalah 0.55.
- Matriks Konfusi:

Model berhasil memprediksi sebagian besar kelas 0 (1924 prediksi benar, 59 prediksi salah).
Model memiliki kesulitan dalam memprediksi kelas 1 (296 prediksi benar, 221 prediksi salah).
- Akurasi:

Akurasi model adalah 0.858, yang berarti model ini memprediksi kelas dengan benar sekitar 85.8% dari waktu.

## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

In [39]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

# Inisialisasi model KNN
model3 = KNeighborsClassifier()

# Tentukan hyperparameter yang akan di-tune untuk KNN
param_grid = {
    # Jumlah tetangga yang akan dipertimbangkan
    'n_neighbors': [3, 5, 7, 9],
    # Metrik jarak yang akan digunakan. Dua metrik sedang diuji yaitu Euclidean dan Manhattan.
    'metric': ['euclidean', 'manhattan']
}

# Buat objek GridSearchCV untuk KNN
grid = GridSearchCV(
    estimator=model3,
    param_grid=param_grid,
    scoring='accuracy',
    n_jobs=10,
    cv=10
)

# Fit model KNN dan temukan parameter terbaik
grid.fit(X_train, y_train)
best_params = grid.best_params_
print("Parameter terbaik untuk KNN:", best_params)


Parameter terbaik untuk KNN: {'metric': 'manhattan', 'n_neighbors': 5}


In [40]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.85      0.95      0.90      1983
           1       0.63      0.35      0.45       517

    accuracy                           0.82      2500
   macro avg       0.74      0.65      0.67      2500
weighted avg       0.80      0.82      0.80      2500


[[1877  106]
 [ 334  183]]

0.824


**Membaca Hasil Output :**
- Precision:

Kelas 0 (Tidak Ada): Dari semua prediksi kelas 0, sekitar 85% benar (precision 0.85).
Kelas 1 (Ada): Dari semua prediksi kelas 1, sekitar 63% benar (precision 0.63).
- Recall:

Kelas 0 (Tidak Ada): Dari semua instance yang sebenarnya termasuk dalam kelas 0, model dapat memprediksi sekitar 96% dengan benar (recall 0.95).
Kelas 1 (Ada): Dari semua instance yang sebenarnya termasuk dalam kelas 1, model hanya dapat mendeteksi sekitar 35% (recall 0.35).
- F1-Score:

F1-score mencoba mencari keseimbangan antara precision dan recall. Untuk kelas 0, f1-score adalah 0.90, sedangkan untuk kelas 1, f1-score adalah 0.45.
- Matriks Konfusi:

Model berhasil memprediksi sebagian besar kelas 0 (1877 prediksi benar, 106 prediksi salah).
Model memiliki kesulitan dalam memprediksi kelas 1 (334 prediksi benar, 183 prediksi salah).
- Akurasi:

Akurasi model adalah 0.824, yang berarti model ini memprediksi kelas dengan benar sekitar 82.4% dari waktu.

## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

Dalam percobaan ini, tiga model berbeda digunakan yaitu Decision tree, SVM , dan KNN. Hasil akurasi yang diperoleh untuk masing-masing model adalah berbeda, yaitu Decision Tree sebesar 85%, SVM sebesar 86%, dan KNN sebesar 82%. Setelah mengevaluasi model menggunakan metrik evaluasi seperti precision, recall, dan f1-score, serta menganalisis matriks konfusi dan akurasi, dapat disimpulkan bahwa model terbaik di antara ketiganya adalah model SVM dengan akurasi 86%.
